samples/zello_events_deadlock/zello_events_deadlock.cpp

```diff
@@ -0,0 +1,266 @@
+    ze_context_handle_t context;
+    ze_context_desc_t context_desc = {};
+    context_desc.stype = ZE_STRUCTURE_TYPE_CONTEXT_DESC;
+    status = zeContextCreate(pDriver, &context_desc, &context);
+    if (status != ZE_RESULT_SUCCESS) {
+        std::cout << "zeContextCreate Failed with return code: " << to_string(status) << std::endl;
+        exit(1);
+    }
+
+    // Create an immediate command list for direct submission
+    ze_command_queue_desc_t altdesc = {};
+    altdesc.stype = ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC;
+    ze_command_list_handle_t command_list = {};
+    status = zeCommandListCreateImmediate(context, pDevice, &altdesc, &command_list);
+    if (status != ZE_RESULT_SUCCESS) {
+        std::cout << "zeCommandListCreateImmediate Failed with return code: " << to_string(status) << std::endl;
+        exit(1);
+    }
+
+    // Create an event to be signaled by the device
+    ze_event_pool_desc_t ep_desc = {};
+    ep_desc.stype = ZE_STRUCTURE_TYPE_EVENT_POOL_DESC;
+    ep_desc.count = 3;
+    ep_desc.flags = ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+
+    ze_event_pool_handle_t event_pool;
+
+    status = zeEventPoolCreate(context, &ep_desc, 1, &pDevice, &event_pool);
+    if (status != ZE_RESULT_SUCCESS) {
+        std::cout << "zeEventPoolCreate Failed with return code: " << to_string(status) << std::endl;
+        exit(1);
+    }
+
+    std::vector<ze_event_handle_t> event{};
+    // Two events for memcpy that will form a dependency on a 3rd event
+    event.resize(3);
+
+    ze_event_desc_t ev_desc = {};
+    ev_desc.stype = ZE_STRUCTURE_TYPE_EVENT_DESC;
+    ev_desc.signal = ZE_EVENT_SCOPE_FLAG_DEVICE; // ZE_EVENT_SCOPE_FLAG_HOST ??
+    ev_desc.wait = ZE_EVENT_SCOPE_FLAG_HOST;
+
+    ev_desc.index = 0;
+    SUCCESS_OR_TERMINATE(zeEventCreate(event_pool, &ev_desc, &event[0]));
+
+    ev_desc.index++;
+    SUCCESS_OR_TERMINATE(zeEventCreate(event_pool, &ev_desc, &event[1]));
+
+    ev_desc.index++;
+    SUCCESS_OR_TERMINATE(zeEventCreate(event_pool, &ev_desc, &event[2]));
+
+    /* ev_desc.signal = ZE_EVENT_SCOPE_FLAG_HOST;
+    ev_desc.wait = ZE_EVENT_SCOPE_FLAG_DEVICE;
+    ev_desc.index++;
+    ze_event_handle_t start_event;
+    SUCCESS_OR_TERMINATE(zeEventCreate(event_pool, &ev_desc, &start_event)); */
+
+    ze_host_mem_alloc_desc_t host_desc = {};
+    host_desc.stype = ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC;
+    host_desc.pNext = nullptr;
+    host_desc.flags = 0;
+
+    size_t buffer_size = 1024;
+    void *host_mem_ptr = nullptr;
+    SUCCESS_OR_TERMINATE(zeMemAllocHost(context, &host_desc, buffer_size, 1, &host_mem_ptr));
+
+    ze_device_mem_alloc_desc_t device_desc = {};
+    device_desc.stype = ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC;
+    device_desc.pNext = nullptr;
+    device_desc.ordinal = 0;
+    device_desc.flags = 0;
+
+    void *device_mem_ptr = nullptr;
+    SUCCESS_OR_TERMINATE(zeMemAllocDevice(context, &device_desc, buffer_size, 0, pDevice, &device_mem_ptr));
+
+    std::cout << "\n\n\n";
+
```

Comment:
There is an api `zeEventHostSignal` that allows signalling events from host.   If that API were called right here for event[2], then everything would work.    I think the deadlock detection mechanism needs to account for this API in some way

Reply:
But even in that case won't it be an ill formed use of the L0 API? Like a single action is being signaled twice?